KNOX-3048 - Add support for group-based impersonation

moresandeep opened the pull request:
## What changes were proposed in this pull request?

Added support for Group Based impersonation provider. 

```
<!-- turn the feature on -->
<param>
     <name>hadoop.proxygroup.impersonation.enabled</name>
     <value>true</value>
</param>

<!-- proxy group configs -->
<param>
     <name>hadoop.proxygroup.KNOX_ADMIN.groups</name>
     <value>*</value>
</param>

<param>
     <name>hadoop.proxygroup.VIRTUAL_GROUP.groups</name>
     <value>*</value>
</param>

<param>
     <name>hadoop.proxygroup.VIRTUAL_GROUP.hosts</name>
     <value>*</value>
</param>

``` 


## How was this patch tested?
Unit tests and local testing

moresandeep commented:
Folks apology for formatting :( I accidentally formatted the code as i went along to prevent checkstyle issues and now going back to the old formatting is pain. If anyone has objections I can update the code (might take a while though :( )

moresandeep commented:
> > Folks apology for formatting :( I accidentally formatted the code as i went along to prevent checkstyle issues and now going back to the old formatting is pain. If anyone has objections I can update the code (might take a while though :( )
> 
> Hi, @moresandeep ! It'd be beneficial if you updated the code and include the 'real' changes here. Later on, we can have a formatting PR. This would help reviewers to understand the purpose of this one.

Thanks @smolnar82 I will provide a summary of the changes so it will be easy for folks to review.

moresandeep commented:
Following are the major changes in `CommonIdentityAssertionFilter` class

- Centralized “is impersonation enabled” logic
     •	Old: read IMPERSONATION_ENABLED_PARAM directly from filterConfig, parsed its boolean value inline
	•	New: delegated that check to a new helper 
-  Eliminated duplicate constant
	•	Removed the class-local static final String IMPERSONATION_ENABLED_PARAM = AuthFilterUtils.PROXYUSER_PREFIX + ".impersonation.enabled";
        •	Instead statically import AuthFilterUtils.IMPERSONATION_ENABLED_PARAM (and added GROUP_IMPERSONATION_ENABLED_PARAM for future use)

- Helper methods promoted & grouped
	•	Moved virtualGroupParameterNames(...) and unique(...) up near the top of the class, declared them private static so all utility logic is visible before the lifecycle methods.

- No other behavior changes
	•	All the parsing of advanced principal mappings, virtual-group loading, doFilter flow, etc., is untouched—only how and where the impersonation flag is determined was refactored.

moresandeep commented:
**Note**: _The following summary is generated using co-pilot_

This pull request refactors the `CommonIdentityAssertionFilter` class and its related tests to improve code maintainability, readability, and functionality. Key changes include the introduction of a new method for checking impersonation provider requirements, reorganization of imports, and updates to test cases to align with the refactored implementation.

### Refactoring and Code Simplification:
* Introduced the `shouldAddImpersonationProvider` method to encapsulate logic for determining if the impersonation provider should be added, replacing inline checks. (`[gateway-provider-identity-assertion-common/src/main/java/org/apache/knox/gateway/identityasserter/common/filter/CommonIdentityAssertionFilter.javaR170-R180](diffhunk://#diff-4577bd8f7a7a8aa491f44533f5111e9deb9496962b2059d5e0cbb3f1b59c4bfdR170-R180)`)
* Removed redundant methods `virtualGroupParameterNames` and `unique`, and integrated their functionality directly where needed. (`[[1]](diffhunk://#diff-4577bd8f7a7a8aa491f44533f5111e9deb9496962b2059d5e0cbb3f1b59c4bfdL207-L211)`, `[[2]](diffhunk://#diff-4577bd8f7a7a8aa491f44533f5111e9deb9496962b2059d5e0cbb3f1b59c4bfdL299-R312)`)
* Reorganized imports in `CommonIdentityAssertionFilter` to improve clarity and remove unused dependencies. (`[[1]](diffhunk://#diff-4577bd8f7a7a8aa491f44533f5111e9deb9496962b2059d5e0cbb3f1b59c4bfdL20-L46)`, `[[2]](diffhunk://#diff-4577bd8f7a7a8aa491f44533f5111e9deb9496962b2059d5e0cbb3f1b59c4bfdL65-R90)`)

### Test Case Updates:
* Updated test cases in `CommonIdentityAssertionFilterTest` to use the refactored `shouldAddImpersonationProvider` method and replaced references to the removed `IMPERSONATION_ENABLED_PARAM` constant. (`[[1]](diffhunk://#diff-f1e853cfe04089b2ce1381bc562b0d6e96ace1db5ee2c0085fa1dd8bdaaae7dfL286-R264)`, `[[2]](diffhunk://#diff-f1e853cfe04089b2ce1381bc562b0d6e96ace1db5ee2c0085fa1dd8bdaaae7dfL141-R141)`)
* Reorganized imports in `CommonIdentityAssertionFilterTest` for consistency with the refactored implementation. (`[[1]](diffhunk://#diff-f1e853cfe04089b2ce1381bc562b0d6e96ace1db5ee2c0085fa1dd8bdaaae7dfL20-R40)`, `[[2]](diffhunk://#diff-f1e853cfe04089b2ce1381bc562b0d6e96ace1db5ee2c0085fa1dd8bdaaae7dfL41-R60)`)
* Added expectations for new impersonation-related parameters in `HadoopAuthFilterTest` to ensure compatibility with the updated implementation. (`[[1]](diffhunk://#diff-856be9bd8db07e7ffa4042e3f2943b254b2c6204e1bd7e50c6bfae8fa89e3cf3R148-R149)`, `[[2]](diffhunk://#diff-856be9bd8db07e7ffa4042e3f2943b254b2c6204e1bd7e50c6bfae8fa89e3cf3R214-R215)`, `[[3]](diffhunk://#diff-856be9bd8db07e7ffa4042e3f2943b254b2c6204e1bd7e50c6bfae8fa89e3cf3R266-R267)`, `[[4]](diffhunk://#diff-856be9bd8db07e7ffa4042e3f2943b254b2c6204e1bd7e50c6bfae8fa89e3cf3R329-R330)`, `[[5]](diffhunk://#diff-856be9bd8db07e7ffa4042e3f2943b254b2c6204e1bd7e50c6bfae8fa89e3cf3R392-R393)`)

zeroflag commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java:
```diff
@@ -0,0 +1,252 @@
+            /* authorize proxy user */
+            isProxyUserAuthorized = checkProxyUserAuthorization(user, remoteAddress);
+            /* In case of AND no reason to check for proxy groups if isProxyUserAuthorized=false */
+            if("AND".equalsIgnoreCase(impersonationMode) && !isProxyUserAuthorized) {
```
I would either extract this to a string constant, or introduce an enum.

zeroflag commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java:
```diff
@@ -0,0 +1,252 @@
+            isProxyGroupAuthorized = checkProxyGroupAuthorization(user, remoteAddress);
+        }
+
+        if("AND".equalsIgnoreCase(impersonationMode)) {
```
Same as above.

zeroflag commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java:
```diff
@@ -0,0 +1,252 @@
+            return false;
+        }
+
+        boolean proxyGroupHostFound = false;
```
That could be another candidate to extract into its own method.

moresandeep commented on gateway-provider-identity-assertion-common/src/main/java/org/apache/knox/gateway/identityasserter/common/filter/CommonIdentityAssertionFilter.java:
```diff
@@ -62,286 +35,318 @@
+     * @param filterConfig
+     * @return
+     */
+    boolean shouldAddImpersonationProvider(final FilterConfig filterConfig) {
```
@zeroflag yup!

moresandeep commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/AuthFilterUtils.java:
```diff
@@ -27,227 +39,270 @@
+    public static final String PROXYUSER_PREFIX = "hadoop.proxyuser";
+    public static final String PROXYGROUP_PREFIX = "hadoop.proxygroup";
+    public static final String IMPERSONATION_MODE = "hadoop.impersonation.mode";
+    public static final String DEFAULT_IMPERSONATION_MODE = "OR";
```
ugh, i should have updated my PR description 
```
                 <param>
                    <name>hadoop.impersonation.mode</name>
                    <value>OR</value>
                  </param>
```
This would be an example.

moresandeep commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java:
```diff
@@ -0,0 +1,252 @@
+            Collections.addAll(realUserGroups, user.getRealUser().getGroupNames());
+        }
+
+        boolean proxyGroupFound = false;
```
Yup

moresandeep commented:
@smolnar82 @zeroflag thank you so much for the valuable reviews! I going to make the changes and do some more refactoring in light of changes and some debugging. I will mark the PR as WIP for now and let you know when i have it back up. 
Thank You again!!

zeroflag commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/AuthFilterUtils.java:
```diff
@@ -42,212 +45,307 @@
+     * Check if user or group impersonation is enabled based on filter configuration.
+     *
+     * @param filterConfig The filter configuration
+     * @return A boolean array where the first element indicates if user impersonation is enabled
```
Stale javadoc comment. It no longer returns a boolean array.

zeroflag commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java:
```diff
@@ -0,0 +1,228 @@
+        if (!proxyGroupHostFound) {
+            LOG.failedToImpersonateGroupsFromAddress(realUser.getUserName(), realUserGroups.toString(), user.getUserName(), remoteAddress.toString());
+            throw new AuthorizationException("User: " + realUser.getUserName()
+                    + "with groups " + realUserGroups.toString()
```
Missing space before `"with groups`

lmccay reviewed (commented):
@moresandeep - why add a new provider for this and not just add it to common so that it is available everywhere that extends that?

moresandeep commented:
> @moresandeep - why add a new provider for this and not just add it to common so that it is available everywhere that extends that?

Good point, I thought about that but decided against it in favor of code separation. Since all of the implementations use Common any bug might affect all the providers. Not all providers need this the ones that need this feature are virtual groups (already included in default) and hadoop-group-lookup so i decided to create a new provider that supports these.

Putting this in common will be easy but this is not really common feature that can be used by all the other providers. 
Do you anticipate this to be used by others? i can move this common in that case.

pzampino commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java:
```diff
@@ -0,0 +1,257 @@
+         */
+        if (doesProxyUserConfigExist) {
+            super.authorize(user, remoteAddress);
+        } else{
```
I would expect this to be evaluated in two cases:

1. There is no proxyuser config
2. There is no matching proxyuser config for this specific user

I would not expect this to be an if-else relationship.

pzampino replied:
I mean that the presence of proxyuser config should not preclude the evaluation of the group-level config. If there is user-level config that does not apply to the requesting user (super.authorize() fails), then the group-level config should be evaluated to see if there is matching authorization based on the requesting user's group membership.

pzampino replied:
Probably means catching AuthorizationException from the super.authorize() call, and continuing if there is group-level config. Perhaps, something like
`if(doesProxyUserConfigExist && doesProxyUserConfigExist(user)) {
        try {
            super.authorize(user, remoteAddress);
        } catch (AuthorizationException e) {
            if (!doesProxyGroupConfigExist) {
                throw e;
            }
        }
} 

if (doesProxyGroupConfigExist) {
        checkProxyGroupAuthorization(user, remoteAddress, groups);
}
`

moresandeep replied:
Ah, i see thanks for the explanation!

moresandeep commented:
@pzampino i addressed your comments, following are the changes in a nutshell:
1. Regex expressions are now constants
2.  For cases where proxy user impersonation fails proxy group proxy impersonation will be evaluated
3. Added unit tests for the item no. 2

I tested it locally and verified that it works. Let me know if there are more changes.

pzampino commented on gateway-spi/src/main/java/org/apache/knox/gateway/i18n/GatewaySpiMessages.java:
```diff
@@ -93,4 +96,22 @@ public interface GatewaySpiMessages {
+  @Message(level=MessageLevel.DEBUG, text="Loaded proxy groups ACLs: {0}")
+  void loadedProxyGroupsAcls(String acls);
+
+  @Message(level = MessageLevel.ERROR, text = "Unauthorized connection for super-user: {0} from IP remoteAddress {1}")
```
Suggestion: "User impersonation failed for user {0}. Connections from remote address {1} are not authorized."

pzampino commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java:
```diff
@@ -0,0 +1,263 @@
+            } catch (final AuthorizationException e) {
+                /*
+                 * Log and try group based impersonation.
+                 * Since this provider is for groups no need to check if
```
Is this comment true? We only employ this implementation if there is group-level proxy config? That seems wrong to me.

moresandeep replied:
The main reason was separation of functionality. I did not want the existing functionality to be affected by new code. So, unless users configure proxy group `GroupBasedImpersonationProvider` will not be engaged. This was my thinking.

pzampino commented on gateway-spi/src/main/java/org/apache/knox/gateway/util/AuthFilterUtils.java:
```diff
@@ -42,212 +44,247 @@
-        public Principal getUserPrincipal() {
-          return remoteRequestUgi::getUserName;
+        ImpersonationProvider impersonationProvider;
+        if(hasProxyGroupParams) {
```
Why can't this be our default implementation that supports both user-level and the new group-level impersonation config? It would simplify a lot of things. So, rather than GroupBasedImpersonationProvider, it would be someyhing like KnoxImpersonationProvider (our implementation that also supports group-level config).

pzampino reviewed (commented):
The change you've made to address my previous comment appears to match what I had in mind. I just have a couple more questions.
